packages/hub/src/graphql/queries/search.ts

```diff
@@ -0,0 +1,126 @@
+import { builder } from "@/graphql/builder";
+import { prisma } from "@/prisma";
+import { Searchable } from "../types/Searchable";
+import { Searchable as PrismaSearchable } from "@prisma/client";
+import { resolveOffsetConnection } from "@pothos/plugin-relay";
+
+type SearchableWithEdgeData = PrismaSearchable & {
+  rank: number;
+  slugSnippet: string;
+  textSnippet: string;
+};
+
+const SearchEdge = builder.edgeObject({
+  type: Searchable,
+  name: "SearchEdge",
+  fields: (t) => ({
+    rank: t.float({
+      resolve: (obj) => {
+        const rank = (obj.node as SearchableWithEdgeData).rank;
+        return rank ?? 1e9;
+      },
+    }),
+    slugSnippet: t.string({
+      resolve: (obj) => {
+        const snippet = (obj.node as SearchableWithEdgeData).slugSnippet;
+        return String(snippet ?? "");
+      },
+    }),
+    textSnippet: t.string({
+      resolve: (obj) => {
+        const snippet = (obj.node as SearchableWithEdgeData).textSnippet;
+        return String(snippet ?? "");
+      },
+    }),
+  }),
+});
+
+builder.queryField("search", (t) =>
+  t.connection(
+    {
+      type: Searchable,
+      args: {
+        text: t.arg.string({ required: true }),
+      },
+      errors: {},
+      resolve: async (_, args, { session }) => {
+        return await resolveOffsetConnection(
+          { args },
+          async ({ limit, offset }) => {
+            const { text } = args;
+            return await prisma.$queryRaw<SearchableWithEdgeData[]>`
+        SELECT
+          "Searchable".*,
+          -- ranking function
+          ts_rank_cd(
+            setweight(to_tsvector(coalesce("Model".slug, '')), 'A') ||
+            setweight(to_tsvector(coalesce("SquiggleSnippet".code, '')), 'B') ||
+            setweight(to_tsvector(coalesce("RelativeValuesDefinition".slug, '')), 'A') ||
+            setweight(to_tsvector(coalesce("UserOwner".slug, '')), 'A') ||
+            setweight(to_tsvector(coalesce("GroupOwner".slug, '')), 'A'),
+            websearch_to_tsquery(${text})
+          ) AS rank,
+          ts_headline(
+            concat_ws(
+              ' ',
+              "Model".slug,
+              "RelativeValuesDefinition".slug,
+              "UserOwner".slug,
+              "GroupOwner".slug
+            ),
+            websearch_to_tsquery(${text}),
+            'HighlightAll=t'
+          ) AS "slugSnippet",
```

Comment:
I wanted to do `titleSnippet` (highlight the entire "ownerSlug/slug" string instead of only "slug" part) here, but discovered that Postgres thinks that `foo/bar` is a single token, and customizing the parser is [possible](https://www.postgresql.org/docs/16/sql-createtsparser.html), but [complicated](https://dba.stackexchange.com/a/251976).